release-23.1: kvserver: sync before removing sideloaded files

pav-kv opened the pull request:
Backport:
  * 2/2 commits from "server: support strict MemFS in StickyVFSRegistry" (#115595)
  * 6/6 commits from "kvserver: sync before removing sideloaded files" (#114191)

Please see individual PRs for details.

/cc @cockroachdb/release

Release justification: critical bug fix

pav-kv committed:
kvserver: document state machine sync nuances

Epic: none
Release note: none

pav-kv committed:
server: support strict MemFS in StickyVFSRegistry

The strict MemFS can be used for emulating crashes, and testing for data
durability loss that they may cause if fsync is not used correctly.

Epic: none
Release note: none

pav-kv committed:
logstore: regression test for sideloaded storage panic

This commit adds a regression test for the issue fixed in the previous
commit. Without the commit, the test panics because the sideloaded
storage loses entries which the follower Replica will try to apply after
a restart.

For a detailed explanation of the scenario, see the previous commit and
the comment in the test.

Epic: none
Release note: none

pav-kv committed:
kvserver: sync before removing sideloaded files

A typical raft log truncation command removes log entries and updates
the state machine in a single Pebble batch. This ensures that the state
machine's TruncatedState is always up-to-date with the log.

However, if some of the truncated entries are stored in the sideloaded
storage (typically AddSST commands), the corresponding files are removed
separately from the Pebble batch (when applying side effects). If the
batch is not synced before the files are removed, a process crash may
cause an inconsistent state: the TruncatedState in the state machine
does not reflect the removal of the entries, but the entry files were
removed. After restart, raft may try to load these entries in order to
apply them or send to other replicas. It will find them missing and
crash loop.

This commit syncs the state machine upon applying truncation commands
that remove at least one sideloaded entry, to prevent the issue above.

Epic: none

Release note (bug fix): this commit fixes a bug in raft log truncation
that could lead to crash loops, and unrecoverable loss of quorum in the
unlikely worst case that all replicas enter this crash loop. The bug
manifests when a few things coincide. The cluster is running a bulk
write workload (e.g. schema change, import, restore), a log truncation
command is running, and CRDB process crashes at an unfortunate moment
(e.g. the process is killed, or kills itself for reasons like when it
detects a disk stall).

pav-kv committed:
server: return concrete MemFS from StickyVFSRegistry

The StickyVFSRegistry is only used with in-memory vfs.MemFS for now.
Return the concrete type from the Get method directly, to avoid
unnecessary casts on the caller site.

Epic: none
Release note: none

pav-kv committed:
kvserver: intercept raft sync and post-apply

This commit adds a couple of interceptors into the raft proposal's
lifetime: one allows blocking the command application flow (without
blocking the entry commit flow), another allows intercepting commands
after applying them to storage and applying the in-memory side effects.

Both interceptors are useful in the next commits that reproduce a
problem in raft log's sideloaded storage.

Epic: none
Release note: none

pav-kv committed:
logstore: option to disable log append randomization

This commit adds a knob that allows tests to disable the randomization
which sometimes makes asynchronous log appends synchronous. This is
necessary for tests that rely on the async log appends.

Epic: none
Release note: none

pav-kv committed:
logstore: add HasAnyEntry check to sideloaded storage

To decide whether log truncation requires sync, we need to check whether
there are any sideloaded entries in the removed log entries range. This
commit adds the corresponding helper method to the sideloaded storage.

The existing BytesIfTruncatedFromTo method could be used instead, but it
is more expensive: it fetches stats for all the files it scans, whereas
we just need an existence check, and we can terminate the scan early.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/117299)
<!-- Reviewable:end -->

pav-kv commented:
Depends on #117295

pav-kv commented:
Hm, the test failed with an interesting error:

```
=== RUN   TestCrashWhileTruncatingSideloadedEntries
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestCrashWhileTruncatingSideloadedEntries412629092
    test_log_scope.go:79: use -show-logs to present logs inline
    client_raft_log_queue_test.go:234: leader replica: [n1,s1,r64/1:/{Table/Max-Max}]
    client_raft_log_queue_test.go:239: follower replica: [n2,s2,r64/2:/{Table/Max-Max}]
    client_raft_log_queue_test.go:245: leader: log indices: [11..20]
    client_raft_log_queue_test.go:246: leader: applied to: 20
    client_raft_log_queue_test.go:245: follower: log indices: [17..20]
    client_raft_log_queue_test.go:246: follower: applied to: 20
    client_raft_log_queue_test.go:271: committed AddSSTs
    client_raft_log_queue_test.go:245: leader: log indices: [11..40]
    client_raft_log_queue_test.go:246: leader: applied to: 40
    client_raft_log_queue_test.go:245: follower: log indices: [17..40]
    client_raft_log_queue_test.go:246: follower: applied to: 20
    client_raft_log_queue_test.go:299: committed truncation for indices < 39
    client_raft_log_queue_test.go:245: leader: log indices: [39..41]
    client_raft_log_queue_test.go:246: leader: applied to: 41
    client_raft_log_queue_test.go:245: follower: log indices: [17..41]
    client_raft_log_queue_test.go:246: follower: applied to: 20
    client_raft_log_queue_test.go:319: unblocked follower application flow
    client_raft_log_queue_test.go:321: follower applied the truncation
    client_raft_log_queue_test.go:245: follower: log indices: [39..41]
    client_raft_log_queue_test.go:246: follower: applied to: 41
    client_raft_log_queue_test.go:345: CRASH!
    client_raft_log_queue_test.go:349: restarting follower
    client_raft_log_queue_test.go:352: FS after restart:
                  /
             892    000002.log
           10853    000010.log
            1816    000012.sst
            1163    000013.sst
            1018    000014.sst
            1120    i21.t6
            1120    i22.t6
            1120    i23.t6
            1120    i24.t6
            1120    i25.t6
            1120    i26.t6
            1120    i27.t6
            1120    i28.t6
            1120    i29.t6
            1120    i30.t6
            1120    i31.t6
            1120    i32.t6
            1120    i33.t6
            1120    i34.t6
            1120    i35.t6
            1120    i36.t6
            1120    i37.t6
            1120    i38.t6
            1120    i39.t6
            1120    i40.t6
              16    CURRENT
               0    LOCK
            2226    MANIFEST-000001
            2538    OPTIONS-000003
               8    STORAGE_MIN_VERSION
                    auxiliary/
                      sideloading/
                        r0XXXX/
                          r64/
            1120            i39.t6
            1120            i40.t6
                      sstsnapshot/
               0    marker.format-version.000012.013
               0    marker.manifest.000001.MANIFEST-000001
F240105 18:43:53.285591 4473 3@pebble/internal/base/filenames.go:147  [n2,s2,pebble] 1  000035.sst:
F240105 18:43:53.285591 4473 3@pebble/internal/base/filenames.go:147  [n2,s2,pebble] 1 +orig err: open 000035.sst: file does not exist
```

A pebble file is missing after a crash+restart. I'll investigate, wonder if this could be a Pebble bug.

erikgrinaker commented:
>A pebble file is missing after a crash+restart. I'll investigate, wonder if this could be a Pebble bug.

Are these any of the sideloaded files, or just a "regular" LSM SST? Since we don't sync state application, I suppose it's plausible that we could lose a file that was recently flushed from the memtable, but the data is still there in the WAL (and maybe also in the memtable).

Have we seen any failures like these in the nightlies?

erikgrinaker commented:
Ah nevermind, looks like this is an actual Pebble error, not a failed test assertion.

pav-kv commented:
Hasn't been caught in nightlies, but has occurred twice on 23.1, on this PR.

pav-kv commented:
Fixed the flakiness (required backporting #117523 first).

pav-kv commented:
Now there is a legitimate flake. Looks similar to https://github.com/cockroachdb/cockroach/pull/114191#issuecomment-1857083015.

Maybe the `23.1` circuit breakers don't do such a good job at isolating the node as the new ones in `master`.

erikgrinaker commented:
>Maybe the 23.1 circuit breakers don't do such a good job at isolating the node as the new ones in master.

That's probably true, the circuit breakers use a different implementation in 23.2.

pav-kv committed:
kvserver: skip sideloaded crash test under stress

The test is flaky. Possibly due to the circuit breakers auto-recovering
after being tripped, whereas we need to permanently trip them.

Epic: none
Release note: none

pav-kv commented:
Skipping the test under stress to land this fix. Filed #117785 to fix the flakiness (likely has to do with the circuit breaker behaviours).
